File: 0142-linked-list-cycle-ii/0142-linked-list-cycle-ii.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */

public class Solution {
    public ListNode detectCycle(ListNode head) {
        if (head == null) return null;

        ListNode slow=head;
        ListNode fast=head;

        while(fast!=null && fast.next!=null){
            slow=slow.next;
            fast=fast.next.next;

            if(slow==fast)
            {
                ListNode pointer1=head;
                ListNode pointer2=slow;

                while(pointer1!=pointer2){
                    pointer1=pointer1.next;
                    pointer2=pointer2.next;
                }
                return pointer1; 
            }
      }
      return null;
    }
}

File: 0142-linked-list-cycle-ii/0142-linked-list-cycle-ii.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */

/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var detectCycle = function(head) {
    let slow=head;
    let fast=head;

    while(fast!=null && fast.next!=null){
        slow=slow.next;
        fast=fast.next.next;

        if(slow==fast){
            let firstNode=head;

            while(firstNode!=slow){
                firstNode=firstNode.next;
                slow=slow.next;
            }
            return firstNode;
        }
    }

    return null;
};